Src/tools/metaconfig/jmake/jmake.man
''' $Id: jmake.man 78389 2004-11-30 00:17:17Z manus $
'''
'''  Copyright (c) 1991-1993, Raphael Manfredi
'''  
'''  You may redistribute only under the terms of the Artistic Licence,
'''  as specified in the README file that comes with the distribution.
'''  You may reuse parts of this distribution only within the terms of
'''  that same Artistic Licence; a copy of which may be found at the root
'''  of the source tree for dist 3.0.
'''
''' $Log$
''' Revision 1.1  2004/11/30 00:17:18  manus
''' Initial revision
'''
''' Revision 3.0.1.1  1995/05/12  11:57:58  ram
''' patch54: updated my e-mail address
'''
''' Revision 3.0  1993/08/18  12:04:18  ram
''' Baseline for dist 3.0 netwide release.
'''
.TH JMAKE 1 ram
.SH NAME
jmake \- a generic makefile builder
.SH SYNOPSIS
.B jmake
[
.I cpp options
]
.SH DESCRIPTION
.I Jmake
builds a makefile out of a rather high level description held in a
.I Jmakefile
file. The generated file is a
.I Makefile.SH
rather than a simple makefile, which means it is ready to be used in
conjonction with
.I metaconfig.
In particular, parameters such as "where to install executables" will
be automatically determined by
.I Configure
and only the needed parameters will be taken into account.
.PP
To use
.I jmake
you have to write a
.I Jmakefile
first, which describes the way things are to be built. Your
.I Jmakefile
will be included inside a generic template through the C pre-processor.
This means you may use the usual C /**/ comments, but not the shell # comments.
The C comments will not appear in the generated
.I Makefile.SH
but lines starting with ;# will finally appear as shell comments. If you
have to write the string /* in the generated
.I Makefile.SH
then you have to escape it (otherwise
.I jmake
will think of it as the start of a C comment). Simply put a # in front
of the *, as in /#*.
.PP
You have a set of macros at your disposal, and all these macros are
listed in the Index file, along with the piece of code they will
expand to. Usually, a
.I Jmakefile
is fairly small and thus easier to maintain than a huge
.I Makefile.
Some internal powerful commands allow you to write
portable makefiles easily, without having to spend many efforts, because
someone else already did the job for you :-).
.PP
When you want to generate your makefile, you usually do not run
.I jmake
but use the
.I jmkmf
script which is a wrapper and will invoke
.I jmake
with the correct options.
.PP
All the knowledge of
.I jmake
is held in two files: the template
.I Jmake.tmpl
and the macro definition file
.I Jmake.rules.
The first file includes the second, along with the
.I Jmakefile.
It is sometimes necessary to know how things works to be able to correctly
use all the features provided. For instance, you may have to write your
own rules for a specific project. Although you cannot overwrite the
predefined rules, you can extent the
.I Jmake.rules
file or simply add your macros in your
.I Jmakefile.
You may also use
.I #include
statements when you want to share these macros and do not want to duplicate
the code.
.PP
The syntax in Jmake.rules is not elegant at all, but:
.sp
.PD 0
.IP -
It is easy to parse (like sendmail.cf or troff files).
.IP -
The rules are not supposed to change very often.
.IP -
It is simple enough to be mastered in five minutes. :-)
.sp
Here is a small description:
.sp
.IP 1)
To deal with various \fIcpp\fR implementations:
.sp
.RS
.IP \(bu
Final @!\\ means: end of line, next line starts at the left margin.
.IP \(bu
Final @@\\ means: end of line, next line is to be indented by one tab.
.sp
.PP
There should always be one of @!\\ or @@\\ at the end of each line.
The only exception is for macros that are to be used as part of a
rule body (e.g. \fIRemoveTargetProgram\fR). In that case, the first
line (which holds the \fI#define\fR) should end with a single backslash.
.RE
.sp
.IP 2)
Symbol definitions:
.sp
.RS
.IP \(bu
>SYMBOL: defines the symbol.
.IP \(bu
?SYMBOL:<text>: keeps <text> iff SYMBOL is defined.
.IP \(bu
%SYMBOL:<text>: keeps <text> iff SYMBOL is not defined.
.sp
.PP
The ?SYM can be nested (logical AND), as in:
.sp
.in +5
?SYMBOL:%TOKEN:text
.in -5
.sp
which will keep text if SYMBOL is defined and TOKEN undefined.
To implement a logical OR, see below.
.RE
.sp
.IP 3)
Commands:
.sp
.RS
Commands can be passed to \fIjmake\fR. They start with a leading '|'.
Available commands are:
.sp
.IP \(bu
|suffix <sx>: adds <sx> to the .SUFFIXES: list in the makefile.
.IP \(bu
|rule:<text>: adds <text> to the building rule section.
.IP \(bu
|rule: <text>: same as before, with a leading tab.
.IP \(bu
|skip: skips text until a line starting with '-skip' is found.
.IP \(bu
|expand <pattern>: expand lines until '-expand' with <pattern>. A
complete example is shown below.
.IP \(bu
|once <symbol>: text up to '-once' appears only the first time.
.sp
.PP
Here is a way to implement a logical OR:
.sp
.in +5
.nf
/* Implements SYMBOL or not TOKEN */
?SYMBOL:text		/* Keeps text if SYMBOL */
%SYMBOL:|skip
	%TOKEN:text		/* Keeps text if not TOKEN */
-skip
.fi
.in -5
.sp
Actually, this is ugly, because the text has to appear twice.
Fortunately, I did not use it. :-)
.sp
.PP
The '|' commands cannot be nested. In particular, due to the simple
implementation of \fI|skip\fR, it is impossible to put \fI|skip\fR inside
a skipped part. However, a \fI|once\fR section may have \fI|skip\fR sections.
.sp
But actually, as you have surely already guessed, the best way to
implement a logical OR is to use De Morgan's Law:
.sp
.in +5
.nf
not (p or q) <=> not p and not q

/* Implements SYMBOL or not TOKEN (attempt #2) */
%SYMBOL:?TOKEN:|skip
text					/* If SYMBOL or not TOKEN */
-skip
.sp
.in -5
.fi
Who said they don't care ? ;-)
.sp
.PP
Expansion is done with the \fIexpand\fR command.  It has been provided to
avoid some cumbersome writings in makefiles when you have to repeat some
silly lines that only differ in file names, for instance.  Let's look at
an example first:
.sp
.in +5
.nf
|expand a!foo bar! b!yes no!
!a::
	echo !a, !b
-expand
.fi
.in -5
.sp
.PP
Then two rules will be printed, and the values of (a,b) for the first
will be (foo, yes), for the second (bar, no).  Substitution is controled
by the '!' character.  If the word to be substituted is part of another
one, detach with the ^^ construct as in:  !b^^c.  It is possible to
use Makefile macros in the <pattern>, and they will be expanded by
jmake.  If this is not what you want, escape the first '$' sign (this is
a Makefile escape, i.e. you must double the '$', not precede it with a
backslash). A // stands for the null substitution value.
.sp
.PP
Here is another example which shows how the macro Expand can be used.
It is defined in \fIJmake.rules\fR as:
.sp
.in +5
.nf
#define Expand(rule, pattern) @!\\
|expand pattern @!\\
rule @!\\
-expand
.sp
.in -5
.fi
So we can write in the \fIJmakefile\fR:
.sp
.in +5
.nf
|skip
A = foo bar
-skip

#define Rule @!\\
$(DIR)/!a^^.o: !a^^.o @@\\
	$(CC) -c !a^^.c @@\\
	$(MV) !a^^.o $(DIR)

Expand(Rule, a!$(A)!)
.sp
.in -5
.fi
which will generate in \fIMakefile.SH\fR:
.sp
.in +5
.nf
$(DIR)/foo.o: foo.o
	$(CC) -c foo.c
	$(MV) foo.o $(DIR)

$(DIR)/bar.o: bar.o
	$(CC) -c bar.c
	$(MV) bar.o $$(DIR)
.sp
.in -5
.fi
.sp
.PP
The 'A' declaration has been surrounded by \fIskip\fR, so that it does
not appear in the generated Makefile.SH, but it will be taken into
account by \fIjmake\fR for the substitution in the pattern.
.sp
.PP
The number of expansions is determined by the number of possible
values for the \fBfirst\fR parameter. If other parameters have less
substitution values, they will get void ones.
.sp
.PP
It is possible to add a regular expression at the end of '-expand'. This
regular expression will be removed from the final set of expansion at the
end of each line. It is also possible to do substitutions in the expanded
item, by using the syntax (if 'f' is the expanded variable)
!f:\fI<p>\fR=\fI<q>\fR
where \fI<p>\fR and \fI<q>\fR are two regular expressions (without spaces).
The pattern \fI<p>\fR will be replaced by the pattern \fI<q>\fR (only the first
occurrence will be replaced).
.sp
.PP
Finally, you may refer in the expanded section to variables whose value is
computed via another expansion, which makes it easy to define generic
\fIJmakefiles\fR.
.sp
Example:
.sp
.in +5
.nf
SRC = foo.c bar.c
OBJ = \\
|expand f!$(SRC)!
	!f:\\.c=\\.o \\
-expand \\\\
INC = \\
|expand f!$(OBJ)!
	!f:\\.o=\\.h \\
-expand \\\\
.fi
.in -5
.sp
which will generate in \fIMakefile.SH\fR:
.sp
.in +5
.nf
SRC = foo.c bar.c
OBJ = \\
	foo.o \\
	bar.o
INC = \\
	foo.h \\
	bar.h
.in -5
.fi
.sp
Do not forget to protect special characters in your regular expressions such
as backslash, point, etc...
.sp
.PP
The \fIonce\fR command is tagged with a name. The first time the name
appears, the once construct is ignored and the text up to '-once' will
be copied in the generated Makefile.SH.  However, future occurences of
the same name will be ignored (\fIonce\fR will behave like \fIskip\fR).
.sp
Example:
.sp
.in +5
.nf
|once this_is_a_name
<text>
-once
.sp
.in -5
.fi
.sp
.RE
.IP 4)
Initializations:
.sp
.RS
.IP \(bu
+<line>: Puts the whole line in the initialization section.
.IP \(bu
++SYMBOL <value>: Adds <value> to the SYMBOL macro.
.RE
.sp
.IP 5)
User-defined variables:
.sp
The user may define CFLAGS, LDFLAGS or DPFLAGS as additional flags to be used
in C compilation, linking phase or depend target. It is thus possible to add
some extra flags such as -I or libraries for Makefiles in specific
sub-directories.
.sp
.PD
.SH AUTHOR
Raphael Manfredi <ram@hptnos02.grenoble.hp.com>
.SH FILES
.PD 0
.TP 20
Jmakefile
High level description of Makefile.SH
.TP
Jmake.rules
File holding the macro definitions
.TP
Jmake.tmpl
Template used to mould Makefile.SH
.PD
.SH BUGS
On systems whose
.I cpp
reduces multiple tabs and spaces to a single space,
.I jmake
attempts to put back any necessary tabs (which
.I make
expects in front of rules) but does not properly formats the
body of the rule itself.
.PP
There is a bootstraping problem when creating the first Makefile.SH, because
you cannot run it through a shell until there is a decent Configure
script, but you can't run \fImetaconfig\fR before there is a Makefile.SH
or some needed symbols will not be defined.
.SH "SEE ALSO"
jmkmf(1), metaconfig(1).


Src/tools/metaconfig/jmake/jmake.SH
case $CONFIG in
'')
	if test -f config.sh; then TOP=.;
	elif test -f ../config.sh; then TOP=..;
	elif test -f ../../config.sh; then TOP=../..;
	elif test -f ../../../config.sh; then TOP=../../..;
	elif test -f ../../../../config.sh; then TOP=../../../..;
	else
		echo "Can't find config.sh."; exit 1
	fi
	. $TOP/config.sh
	;;
esac
case "$0" in
*/*) cd `expr X$0 : 'X\(.*\)/'` ;;
esac
echo "Extracting jmake/jmake (with variable substitutions)"
$spitshell >jmake <<!GROK!THIS!
$startperl
	eval 'exec perl -S \$0 "\$@"'
		if \$runnning_under_some_shell;

# $Id: jmake.SH 78389 2004-11-30 00:17:17Z manus $
#
#  Copyright (c) 1991-1993, Raphael Manfredi
#  
#  You may redistribute only under the terms of the Artistic Licence,
#  as specified in the README file that comes with the distribution.
#  You may reuse parts of this distribution only within the terms of
#  that same Artistic Licence; a copy of which may be found at the root
#  of the source tree for dist 3.0.
#
# $Log$
# Revision 1.1  2004/11/30 00:17:18  manus
# Initial revision
#
# Revision 3.0.1.6  1995/09/25  09:08:01  ram
# patch59: will now force macro definitions to the left
#
# Revision 3.0.1.5  1995/07/25  13:34:47  ram
# patch56: all error messages are now prefixed with the program name
#
# Revision 3.0.1.4  1995/03/21  08:45:27  ram
# patch52: now invokes cpp through new fixcpp script
# patch52: first pass now skips cpp comments alltogether
#
# Revision 3.0.1.3  1994/10/29  15:47:01  ram
# patch36: added various escapes in strings for perl5 support
#
# Revision 3.0.1.2  1993/08/24  12:12:50  ram
# patch3: privlib dir was ~name expanded in the wrong place
#
# Revision 3.0.1.1  1993/08/19  06:42:13  ram
# patch1: leading config.sh searching was not aborting properly
#
# Revision 3.0  1993/08/18  12:04:17  ram
# Baseline for dist 3.0 netwide release.
#

\$dir = '$privlib/files';
\$cpp = '$cpp';
\$version = '$VERSION';
\$patchlevel = '$PATCHLEVEL';
!GROK!THIS!
$spitshell >>jmake <<'!NO!SUBS!'

($me = $0) =~ s|.*/(.*)|$1|;
$dir = &tilda_expand($dir);		# ~name expansion
$file = $dir . '/Jmake.tmpl';

$cpp_opt = "-I. ";				# For Jmakefile, which is local
while ($ARGV[0] =~ /^-/) {
	$_ = shift;
	last if /--/;
	$cpp_opt .= "$_ ";
}
$cpp_opt .= "-I$dir";

# Thank you HP-UX for having a cpp that blindly strips trailing backslashes
# in the text. Run through cpp by using the fixcpp script...

open(CPP, "$dir/fixcpp $cpp_opt $file |");
while (<CPP>) {
	# Record defined symbols in Jmakefile. We won't catch symbols
	# in conditional commands, but that's ok, I hope.
	if ($in_symbol) {
		$val = $_;
		$in_symbol = 0 if !($val =~ s/\\\s*$//);	# Last line
		if ($val = /^\|expand/) {		# Found an expand command
			$in_symbol = 0;				# Stop gathering value
			$val .= "void::x";			# Stop any incomplete escape sequence
		}
		chop($val);
		$Makesym{$current_symbol} .= $val;
	} elsif (/^\s*(\w+)\s*=(.*)/ && !$in_symbol) {
		# Found a makefile's macro declaration
		$val = $2;
		$current_symbol = $1;
		if ($val =~ s/\\\s*$//) {	# Remove final '\'
			$in_symbol = 1;			# This is a continuation line
		}
		$Makesym{$current_symbol} = $val;
		push(@Order, $current_symbol);	# Keep track of order
	}
	# Protect RCS keyword Id or Header from normal substitution
	s/\$(Id|Header|Log)/\$X-$1/;
	# Restore possibly escaped C comments
	s|/#\*|/*|g;
	s|\*#/|*/|g;
	# Remove all ^^ (null space character)
	s|\^\^||g;
	# Restore escaped ^^ sequence
	s|\^\\\^|^^|g;
	next if /^#\s+\d+/;		# Skip cpp commments

	s/^;#/#/;
	s/@#\s?/\n/g;		# Kept for backward compatibility
	s/@!\s?/\n/g;
	s/@@\s?/\n\t/g;

	$* = 1;
	# A '\r' is added to all lines, in order to let 'split' keep them
	# As lines ending with '\' won't fit in the next regular
	# expression (why ?), we have to treat that special case separately
	s/\n$/\r\n/g;
	s/\\\s*$/\\\r/g;	# Remove spaces after final '\' and add '\r'
	$* = 0;
	@macro = split(/\n/);
	for ($i = 0; $i <= $#macro; $i++) {
		chop($_ = $macro[$i]);				# Remove final '\r'
		s/\s+$//g;							# Remove possible useless spaces
		if (/^TOP\s*=\s*(\S+)/) {			# Get the top from generated file
			$top = $1;
		}
		if (s/^\s*>//) {					# '>' means "symbol wanted"
			$symbol{$_} = 1;
		} elsif (s/^\s*\+//) {				# '+' means "initialization section"
			if (s/^\+(\w+)//) {				# '++' means add to variable list
				$added{$1} .= $_;
			} else {						# A single '+' means "add as is".
				push(@init, $_);
			}
		} elsif (s/^\|//) {					# Command for us
			if (/suffix\s+(\S+)/) {			# Add suffix
				push(@suffix, $1) unless $seen{$1};
				$seen{$1} = 1;
			} elsif (s/^rule://) {			# Add building rule
				s/^\s(\s*\S+)/\t$1/;		# Make sure leading tab is there
				push(@rule, $_);
			} elsif (/^skip/) {				# Unconditional skip... funny!
				push(@makefile, "|$_");		# Skip handled in pass 2
			} elsif (/^expand/) {
				push(@makefile, "|$_");		# Expand handled in pass 2
			} elsif (/^once\s+(\S+)/) {		# Once handled in pass 1
				if ($Once{$1}++) {			# Symbol already seen -- skip
					for (; $i <= $#macro; $i++) {
						last if $macro[$i] =~/^-once/;
					}
					warn("$me: -once not found for $1")
						unless $macro[$i] =~/^-once/;
				}
			} else {
				print "$me: Warning: unknown command $_\n";
			}
		} else {
			next if /^-once/;			# Control statement removed
			push(@makefile, $_);
		}
	}
}
close CPP;

@key = keys(%added);
$last_was_blank = 1;	# To avoid blank line at the top of the file
$symbol{'INIT'} = 1 if ($#init >= 0 || $#key >=0);		# Initializations
$symbol{'SUFFIX'} = 1 if ($#suffix >= 0 || $#rule >=0);	# Rules or suffixes
$symbol{'TOP'} = 1 if $top eq '.';		# If imake invoked for the top

open(MAKEFILE, ">Makefile.SH");
# We have to use for instead of foreach to handle 'skip' easily
line: for ($i = 0; $i <= $#makefile; $i++) {
	$_ = $makefile[$i];
	next if /^-skip|-expand/;		# They might have made a mistake

	s/<TAG>/[jmake $version PL$patchlevel]/;
	if (/^\s*$/) {
		next if ($last_was_blank);
		$last_was_blank = 1;
	} else {
		$last_was_blank = 0;
	}

	# Lines starting with ?SYMBOL: (resp. %SYMBOL:) are to be processed
	# only if SYMBOL is defined (resp. undefined).

	# Apply in sequence
	while (/^\s*\?|\s*%/) {
		if (s/^\s*\?(\w+)://) {					# Wanted symbol ?
			next line unless $symbol{$1};
		} elsif (s/^\s*%(\w+)://) {				# Unwanted symbol ?
			next line if $symbol{$1};
		} else {
			print "$me: Warning: missing ':' in $_\n";
			last;
		}
	}

	# We wish to make sure there is a leading tab if the line starts with
	# a space to prevent problems later on. However, variable definitions
	# might want to be aligned on the '=' (imake style). Not all make
	# may be able to cope with those though, so they are left justified
	# again.

	s/^\s/\t/ unless /^\s+\w+\s+=/;		# Make sure leading tab is there
	s/^\s+(\w+\s+=)/$1/;				# Left justify variable definition
	s/^;#/#/;							# Comments in Jmakefile

	if (s/^\|//) {							# Command for us
		if (/^skip/) {						# Skip until -skip
			for (; $i <= $#makefile; $i++) {
				last if $makefile[$i] =~ /^-skip/;
			}
		} elsif (s/^expand//) {
			do init_expand($_);			# Initializes data structures
			$i++;						# Skip expand line
			undef @Expand;				# Storage for expanded lines
			$pattern = '';				# Assume no pattern
			for (; $i <= $#makefile; $i++) {
				$_ = $makefile[$i];
				if (s/^-expand//) {			# Reached end of expansion
					if (s/^\s*(.*)/$1/) {	# Expand followed by a pattern
						$pattern = $_;		# Get pattern to be removed
					}
					last;
				}
				s/^\s/\t/;				# Make sure leading tab is there
				push(@Expand, $_);		# Line to be expanded
			}
			do expand($pattern);		# Expand all lines in buffer
		} else {
			print "$me: Warning: unknown command $_\n";
		}
	} elsif (/^INIT/) {						# Initialization section
		# All the initializations are put in the variable substitution
		# section of the Makefile.SH. Therefore, we have to protect all
		# the '$' signs that are not followed by an alphanumeric character.
		foreach (@init) {
			# Dumps core sometimes with perl 4.0 PL10
			# do protect_dollars(*_);
			$_ = do protect_dollars($_);
			do print_makefile($_);
		}
		foreach (@key) {	# @key set earlier to keys(%added)
			$_ .= " = " . $added{$_};
			# Dumps core sometimes with perl 4.0 PL10
			# do protect_dollars(*_);
			$_ = do protect_dollars($_);
			do print_makefile($_);
		}
	} elsif (/^SUFFIX/) {					# Suffixes/Rules section
		# Rules and suffixes are put in the variable substitution
		# section of the Makefile.SH. Therefore, we have to protect all
		# the '$' signs that are not followed by an alphanumeric character.
		if ($#suffix >= 0) {
			print MAKEFILE ".SUFFIXES:";
			foreach (@suffix) {
				# Dumps core sometimes with perl 4.0 PL10
				# do protect_dollars(*_);
				$_ = do protect_dollars($_);
				print MAKEFILE " $_";
			}
			print MAKEFILE "\n\n";
		}
		foreach (@rule) {
			# Dumps core sometimes with perl 4.0 PL10
			# do protect_dollars(*_);
			$_ = do protect_dollars($_);
			print MAKEFILE "$_\n";
		}
	} else {
		do print_makefile($_);
	}
}
close MAKEFILE;

sub protect_dollars {
	# Dumps core sometimes with perl 4.0 PL10
	# local(*_) = shift(@_);
	s/\\\$/\\=/g;		# Protect already escaped '$'
	s/(\$\W)/\\$1/g;	# Escape unprotected '$'
	s/\\=/\\\$/g;		# Restore escaped '$'
	$_;					# Because perl dumps core... :-(
}

# Initializes data structures for expansion. If we detect Makefile
# macro in the 'expand' line (the argument), then we write a small
# makefile that will do the substitution for us -- I'm lazy today :-)
sub init_expand {
	local($_) = shift(@_);
	undef %Vars;		# Reset array of variables
	$Vars_len = 0;		# Number of "symbols" in first expanded
	if (/\$\(\w+\)/) {	# If at least one macro
		local($make) = "/tmp/mkjm$$";
		open(MAKE, ">$make") || die "$me: can't create $make: $!\n";
		do gen_variables();			# Generates already computed variables
		foreach $var (@Order) {		# Print each in order we found them
			print MAKE "$var = $Makesym{$var}\n" if !$Gvars{$var};
		}
		print MAKE "all:\n\t\@echo '$_'\n";
		close MAKE;
		chop($_ = `make -f $make all`);
		unlink($make);
	}
	while (s/^\s*(\w+)!([^!]*)!//) {
		$Vars{$1} = $2;
		# Record only length for _first_ expanded symbol
		$Vars_len = split(/\s\s*/, $2) unless $Vars_len;
	}
}

# Expand lines in the @Expand array. The argument is a pattern which is to
# be removed from the last chunk of expanded lines.
# For each symbol s, !s is replaced by the next item, and !s:p=q does the
# same after having replaced the pattern 'p' by pattern 'q' in the item.
# Spaces are NOT allowed in 'p' or 'q'. Substitution is done once (no /g).
sub expand {
	local($pattern) = shift;		# To-be-removed pattern for last chunk
	local($_);
	local($sub);
	local($i);
	local(@expands);
	for ($i = 0; $i < $Vars_len; $i++) {
		foreach $line (@Expand) {
			$_ = $line;		# Don't modify elements in array
			foreach $sym (keys %Vars) {
				@expands = split(/\s\s*/, $Vars{$sym});
				$sub = $expands[$i];
				$sub =~ s/\/\///g;		# // is a void value
				while (s/!${sym}:([^\s]*)=([^\s]*)/,x##x,/) {
					# Replacing item is altered by some pattern
					local($p) = $1;
					local($q) = $2;
					local($subq) = $sub;
					eval "\$subq =~ s=${p}=${q}=";
					s/,x##x,/${subq}/;
				}
				s/!${sym}/${sub}/g;
			}
			# Protect substitution in an 'eval' in case of error
			eval "s/${pattern}\$//" if $pattern && $i == ($Vars_len - 1);
			do print_makefile($_);
		}
	}
}

# Prints its argument in MAKEFILE and records it also in Generated
sub print_makefile {
	local($_) = shift(@_);		# Line to be printed
	print MAKEFILE "$_\n";
	push(@Generated, "$_\n");
}

# Generates in MAKE file all the generated variable we have so far for
# final Makefile. This is mainly intended to allow expansion of variables
# which are already defined with an expand.
sub gen_variables {
	undef %Gvars;				# Reset already generated variables
	local ($in_symbol) = 0;		# True when in variable (Makefile's macro)
	foreach (@Generated) {
		if ($in_symbol) {
			if (/^\s*(\w+)\s*=(.*)/) {		# Missed the end of previous macro
				$in_symbol = 0;
				$Gvars{$1} = 1;					# Definition of variable seen
				$in_symbol = 1 if (/\\\s*$/);	# There is a final '\'
				print MAKE "void::\n";			# Cut incomplete esc sequence
			} else  {
				$in_symbol = 0 if !(/\\\s*$/);	# Last line
			}
			print MAKE;
		} elsif (/^\s*(\w+)\s*=(.*)/ && !$in_symbol) {
			# Found a makefile's macro declaration
			$Gvars{$1} = 1;					# Definition of variable seen
			$in_symbol = 1 if (/\\\s*$/);	# There is a final '\'
			print MAKE;
		}
	}
	print MAKE "void::\n";		# Cut incomplete escape sequence
}

!NO!SUBS!
$grep -v '^;#' ../pl/tilde.pl >>jmake 
chmod 755 jmake
$eunicefix jmake
